src/engine/enginemaster.cpp

```diff
@@ -366,16 +406,80 @@ void EngineMaster::process(const int iBufferSize) {
     // Mix all the PFL enabled channels together.
     m_headphoneGain.setGain(chead_gain);
 
-    if (m_bRampingGain) {
-        ChannelMixer::mixChannelsRamping(
-                m_headphoneGain, &m_activeHeadphoneChannels,
-                &m_channelHeadphoneGainCache,
-                m_pHead, iBufferSize);
+    // Mixxx gets input signals delayed by the input latency,
+    // so to preserve the relative timing of the input with the master and headphones
+    // outputs, delay the master, booth, and headphones outputs by the input latency.
+    // This way, if the microphone user is on beat with the output they hear coming from
+    // the speakers/headphones, the input will remain on beat at the expense of
+    // adding some latency.
+    // If the input signal is being directly monitored (mixed with the master output
+    // in hardware), do not delay the master output. The recording/broadcasting signal
+    // will be delayed in this case to keep the recorded/broadcasted signal on beat
+    // and consistent with what the user hears out the speakers/headphones.
+    // If no microphone inputs are configured, avoid adding this extra latency.
+    TalkoverMixMode configuredTalkoverMixMode = static_cast<TalkoverMixMode>(
+        static_cast<int>(m_pTalkoverMixMode->get()));
+    // These latency values are in milliseconds as required by EngineDelay.
+    double inputLatencyCompensation;
+    double measuredRoundTripLatency = m_pRoundTripLatency->get();
```

Comment:
> I do not like this CO, because it can be only a rough assumption.

Yes, it is a rough assumption, but it is better than nothing. Being a few milliseconds off beat is much better than being 10s of millseconds off beat.

> And it does not take the time into account the input monitor needs, or in case of mic the possible extra hardware delay in a pramp or such.

What do you mean by input monitor needs? Do you mean a booth delay? That is a separate use case independent from latency compensation for the master/booth mix relative to mic inputs.

> The idea to help the user by such a value is not bad, but I would move the logic to the preferences as a kind wizzard dialog. Here in the engine it is more important to be able to set directly the final "inputLatency"

I disagree. This is way too complicated to be exposed to users and users do not need to understand it. All they need to do is measure the real roundtrip latency with a loopback cable and Mixxx should do the math correctly.